app/controllers/users/sessions_controller.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class Users::SessionsController < Devise::SessionsController
+  before_action :configure_sign_in_params, only: [:create]
+
+  # GET /resource/sign_in
+  # def new
+  #   super
+  # end
+
+  # POST /resource/sign_in
+  def create
+    # This is abusing REST a bit, but we use a POST for the button
+    # on the homepage to navigate to the log in screen, so we can
+    # pass in the parties, and store them in the session, like we do
+    # with Facebook and Twitter. If we don't have data, they've
+    # probably come from the nav bar, and want to log in. If they do
+    # have the parties, they've probably come from the home page and
+    # want to register, after stashing their parties on the session.
+    # Otherwise it's a log-in request, and we can delegate to Devise.
+    if !params[:user]
+      redirect_to new_user_session_path
+    elsif params[:user][:preferred_party_id] && params[:user][:willing_party_id]
```

Comment:
No objection to REST abuse, but would this be || rather than &&? I think we would still store even if they only set one of the parties?